rust/datafusion/src/datasource/parquet.rs

```diff
@@ -0,0 +1,566 @@
+                            )?
+                        }
+                        ColumnReader::Int96ColumnReader(ref mut r) => {
+                            let mut read_buffer: Vec<Int96> =
+                                vec![Int96::new(); self.batch_size];
+
+                            let mut def_levels: Vec<i16> = vec![0; self.batch_size];
+                            let (_, levels_read) = r.read_batch(
+                                self.batch_size,
+                                Some(&mut def_levels),
+                                None,
+                                &mut read_buffer,
+                            )?;
+
+                            let mut builder = Int64Builder::new(levels_read);
+                            let mut value_index = 0;
+                            for i in 0..levels_read {
+                                if def_levels[i] > 0 {
+                                    builder.append_value(convert_int96_timestamp(
+                                        read_buffer[value_index].data(),
+                                    ))?;
+                                    value_index += 1;
+                                } else {
+                                    builder.append_null()?;
+                                }
+                            }
+                            Arc::new(builder.finish())
+                        }
+                        ColumnReader::FloatColumnReader(ref mut r) => {
+                            ArrowReader::<Float32Type>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::DoubleColumnReader(ref mut r) => {
+                            ArrowReader::<Float64Type>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::FixedLenByteArrayColumnReader(ref mut r) => {
+                            read_binary_column!(self, r, i)
+                        }
+                        ColumnReader::ByteArrayColumnReader(ref mut r) => {
+                            read_binary_column!(self, r, i)
+                        }
+                    };
+
+                    batch.push(array);
+                }
+
+                if batch.len() == 0 || batch[0].data().len() == 0 {
+                    Ok(None)
+                } else {
+                    Ok(Some(RecordBatch::try_new(
+                        self.projection_schema.clone(),
+                        batch,
+                    )?))
+                }
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+/// convert a parquet timestamp in nanoseconds to a timestamp with milliseconds
+fn convert_int96_timestamp(v: &[u32]) -> i64 {
```

Comment:
As `DataType::Timestamp` supports nanosecond precision, might it not be better to keep the resolution at nanosecond level and use a `TimestampNanosecond`? @xhochy do we have to worry about `int96` conversion semantics from https://github.com/apache/parquet-format/pull/49?

Reply:
@andygrove I'll submit another PR that addresses this, please see (https://gist.github.com/nevi-me/574038fdec8e9c207f661813789d58fb)

Reply:
Here is the fix: https://github.com/andygrove/arrow/pull/2

Reply:
Yes, INT96 is that weird. Please ensure that you don't write them by default ;)

Reply:
Thanks @nevi-me and @xhochy !